consensus/types/src/beacon_state.rs

```diff
@@ -191,52 +222,81 @@ where
+    #[superstruct(only(Base))]
     pub previous_epoch_attestations: VariableList<PendingAttestation<T>, T::MaxPendingAttestations>,
+    #[superstruct(only(Base))]
     pub current_epoch_attestations: VariableList<PendingAttestation<T>, T::MaxPendingAttestations>,
 
+    // Participation (Altair and later)
+    #[superstruct(only(Altair))]
+    pub previous_epoch_participation: VariableList<ParticipationFlags, T::ValidatorRegistryLimit>,
+    #[superstruct(only(Altair))]
+    pub current_epoch_participation: VariableList<ParticipationFlags, T::ValidatorRegistryLimit>,
+
     // Finality
     #[test_random(default)]
     pub justification_bits: BitVector<T::JustificationBitsLength>,
+    #[superstruct(getter(copy))]
     pub previous_justified_checkpoint: Checkpoint,
+    #[superstruct(getter(copy))]
     pub current_justified_checkpoint: Checkpoint,
+    #[superstruct(getter(copy))]
     pub finalized_checkpoint: Checkpoint,
 
+    // Inactivity
+    #[superstruct(only(Altair))]
+    pub inactivity_scores: VariableList<u64, T::ValidatorRegistryLimit>,
+
+    // Light-client sync committees
+    #[superstruct(only(Altair))]
+    pub current_sync_committee: SyncCommittee<T>,
+    #[superstruct(only(Altair))]
+    pub next_sync_committee: SyncCommittee<T>,
+
     // Caching (not in the spec)
     #[serde(skip_serializing, skip_deserializing)]
     #[ssz(skip_serializing)]
     #[ssz(skip_deserializing)]
     #[tree_hash(skip_hashing)]
     #[test_random(default)]
+    #[derivative(Clone(clone_with = "clone_default"))]
     pub committee_caches: [CommitteeCache; CACHED_EPOCHS],
     #[serde(skip_serializing, skip_deserializing)]
     #[ssz(skip_serializing)]
     #[ssz(skip_deserializing)]
     #[tree_hash(skip_hashing)]
     #[test_random(default)]
+    #[derivative(Clone(clone_with = "clone_default"))]
     pub pubkey_cache: PubkeyCache,
     #[serde(skip_serializing, skip_deserializing)]
     #[ssz(skip_serializing)]
     #[ssz(skip_deserializing)]
     #[tree_hash(skip_hashing)]
     #[test_random(default)]
+    #[derivative(Clone(clone_with = "clone_default"))]
     pub exit_cache: ExitCache,
     #[serde(skip_serializing, skip_deserializing)]
     #[ssz(skip_serializing)]
     #[ssz(skip_deserializing)]
     #[tree_hash(skip_hashing)]
     #[test_random(default)]
-    pub tree_hash_cache: Option<BeaconTreeHashCache<T>>,
+    #[derivative(Clone(clone_with = "clone_default"))]
+    pub tree_hash_cache: BeaconTreeHashCache<T>,
+}
+
+impl<T: EthSpec> Clone for BeaconState<T> {
+    fn clone(&self) -> Self {
+        self.clone_with(CloneConfig::all())
+    }
 }
 
 impl<T: EthSpec> BeaconState<T> {
     /// Create a new BeaconState suitable for genesis.
     ///
     /// Not a complete genesis state, see `initialize_beacon_state_from_eth1`.
-    ///
-    /// Spec v0.12.1
     pub fn new(genesis_time: u64, eth1_data: Eth1Data, spec: &ChainSpec) -> Self {
-        BeaconState {
+        BeaconState::Base(BeaconStateBase {
```

Comment:
I noticed that `BeaconState` doesn't seem to have a `new` constructor for Altair whilst `BeaconBlock` does:

https://github.com/sigp/lighthouse/blob/1a9c89944dd06b7976a5e10bc4c8697877cea713/consensus/types/src/beacon_block.rs#L60-L66

I'm not far enough in to understand the ramifications, but I'm just making a note of it for now.

Reply:
The reason it's currently like this is that we were creating genesis states for Altair by first making a phase0 state and then upgrading it immediately using `upgrade_to_altair`. This was a bit unwieldy, and the spec has since been updated to clarify genesis initialisation for Altair-only nets: https://github.com/ethereum/eth2.0-specs/pull/2323

When we implement that update from the spec we may well need a `BeaconState::new` that works for Altair